### PR TITLE
feat(models): add quota period database support

### PR DESCRIPTION
**Pull request by michaelbuchar**
Closes https://github.com/reanahub/reana-server/issues/774

### CODE REVIEW

File: reana_db/cli.py [hunk 1]
```diff
@@ -279,15 +278,15 @@ def _resource_usage_update(resource: ResourceType) -> None:
         """Update users resource quota usage."""
         try:
             if resource == ResourceType.disk:
-                update_workflows_disk_quota()
-                update_users_disk_quota()
+                update_workflows_disk_quota(override_policy_checks=True)
```
Comment (michaelbuchar) [inline reana_db/cli.py:281]: Let me know if this is acceptable, I thought you should be able to override policy checks if you want to manually re-trigger the quota `resource-usage-update`

File: reana_db/alembic/versions/20260414_1535_8f2d4e5a4d2c_drop_user_resource_quota_period_anchor.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    column_names = {
+        column["name"]
+        for column in inspector.get_columns("user_resource", schema="__reana")
+    }
+    if "quota_period_anchor_at" in column_names:
+        op.drop_column("user_resource", "quota_period_anchor_at", schema="__reana")
+
+
+def downgrade():
+    """Keep the schema introduced in revision 06dbbeef6d9b."""
```
Comment (tiborsimko) [inline reana_db/alembic/versions/20260414_1535_8f2d4e5a4d2c_drop_user_resource_quota_period_anchor.py]: 🤖 **Empty `downgrade()` makes this migration non-reversible (BLOCKER)**

`upgrade()` drops `quota_period_anchor_at` (line 80), but `downgrade()` is a no-op. If a deployment needs to roll back to the prior schema after spotting a regression, Alembic will report success without restoring the column, and the previous code path that referenced `quota_period_anchor_at` will fail at the next read.

Suggested fix — recreate the column (the original values cannot be recovered, but at least the schema is restored):

```python
def downgrade():
    """Recreate the legacy anchor column (will be NULL for all rows)."""
    op.add_column(
        "user_resource",
        sa.Column("quota_period_anchor_at", sa.DateTime(), nullable=True),
        schema="__reana",
    )
```

**Review by tiborsimko — COMMENTED** [review]
Inline review of the quota period changes. Posting a punch list of items grouped by severity (BLOCKER / HIGH / MEDIUM / LOW / NIT). Happy to discuss any of these.

File: reana_db/alembic/versions/20260414_1535_8f2d4e5a4d2c_drop_user_resource_quota_period_anchor.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                SET quota_period_start_at = :quota_period_start_at
+                WHERE user_id = :user_id
+                  AND resource_id = :resource_id
+                """),
+            {
+                "quota_period_start_at": current_period_start_at,
+                "user_id": row["user_id"],
+                "resource_id": row["resource_id"],
+            },
+        )
```
Comment (tiborsimko) [inline reana_db/alembic/versions/20260414_1535_8f2d4e5a4d2c_drop_user_resource_quota_period_anchor.py]: 🤖 **Per-row `UPDATE` loop will be slow on populated production databases (HIGH)**

The migration issues one `UPDATE` round-trip per affected row. For a deployment with thousands of users (one row per resource per user), this can take a noticeable amount of time and extends the window during which the table is being mutated. The actual computation is a deterministic function of `account_created_at`, `quota_period_months`, and `now()`, so it can be folded into a single SQL statement.

Suggested fix — a pure-SQL `UPDATE ... FROM` using a recursive CTE that walks `account_created_at` forward by `quota_period_months` until it reaches the current window. If we prefer to keep the Python loop for readability, batch the writes into a single `connection.execute(text(...), [...])` call so we make one round-trip instead of N.

If we don't expect populated production databases to have many periodic-quota rows at the time this migration runs, please add a comment in the migration acknowledging that assumption.

File: reana_db/utils.py [hunk 4]
```diff
@@ -491,7 +493,109 @@ def update_workflow_cpu_quota(workflow) -> int:
...
+    if original_period_start_at == current_period_start_at:
+        return False
+
+    user_resource_quota.quota_period_start_at = current_period_start_at
+    logging.info(
+        "Set current CPU quota period for user %s from %s to %s",
+        user_resource_quota.user_id,
+        original_period_start_at,
+        current_period_start_at,
+    )
```
Comment (tiborsimko) [inline reana_db/utils.py:594]: 🤖 **`logging.info` on every period advance will spam logs at scale (MEDIUM)**

`update_users_cpu_quota` runs from a periodic cron job over all users with active tokens. Whenever a period rolls over, this INFO line is emitted. With many users and a short cadence, the signal-to-noise ratio is low — the change is also visible in the database as `quota_period_start_at`, so operators don't need a per-user log entry.

Suggested fix — demote to `logging.debug`:

```python
logging.debug(
    "Set current CPU quota period for user %s from %s to %s",
    ...
)
```

File: reana_db/cli.py [hunk 5]
```diff
@@ -294,3 +300,44 @@ def _resource_usage_update(resource: ResourceType) -> None:
...
+                quota_period_months=DEFAULT_QUOTA_CPU_PERIODIC_RESET_MONTHS,
+            )
+        )
+    Session.commit()
+    updated_rows = len(user_resources)
+
+    click.secho(
+        f"Initialised periodic CPU reset policy for {updated_rows} existing users.",
+        fg="green",
+    )
```
Comment (tiborsimko) [inline reana_db/cli.py]: 🤖 **`init-periodic-defaults` is an unguarded bulk write (MEDIUM)**

The command updates every CPU `UserResource` whose `quota_period_months` is NULL — potentially every existing user — and commits in one go. There is no dry-run preview, no confirmation, and no count printed before the write. Running it on the wrong cluster, or with the wrong default months value in the environment, would be hard to undo.

Suggested fix — add a `--yes` / `--dry-run` pair and print the affected count before committing:

```python
@quota_group.command("init-periodic-defaults")
@click.option("--yes", is_flag=True, help="Do not ask for confirmation.")
@click.option("--dry-run", is_flag=True, help="Report what would change.")
def init_periodic_defaults(yes: bool, dry_run: bool) -> None:
    ...
    click.echo(
        f"Will set periodic CPU quota for {len(user_resources)} users "
        f"({DEFAULT_QUOTA_CPU_PERIODIC_RESET_MONTHS} months)."
    )
    if dry_run:
        return
    if not yes:
        click.confirm("Proceed?", abort=True)
    ...
```

File: reana_db/alembic/versions/20260414_1535_8f2d4e5a4d2c_drop_user_resource_quota_period_anchor.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+def _get_current_quota_period_start_at(account_created_at, quota_period_months):
+    """Derive the active quota window start from the account creation time."""
+    if not account_created_at or not quota_period_months:
+        return None
+
+    now = datetime.utcnow()
+    period_start_at = account_created_at
+    while now >= _add_months(period_start_at, quota_period_months):
+        period_start_at = _add_months(period_start_at, quota_period_months)
+    return period_start_at
```
Comment (tiborsimko) [inline reana_db/alembic/versions/20260414_1535_8f2d4e5a4d2c_drop_user_resource_quota_period_anchor.py]: 🤖 **Period-arithmetic helpers are duplicated between the migration and `utils.py` (LOW)**

`_add_months` and `_get_current_quota_period_start_at` are byte-identical to the equivalents in `reana_db/utils.py:520-541`. Self-containment is the right choice for migrations (they should not import from app code that may evolve), but it does mean a future bug fix to the period arithmetic — for example, the day-clamping at month boundaries — needs to land in two places.

Suggested fix — add a short comment in both copies pointing at the other so this is hard to miss during future maintenance. For example, above the migration helper:

```python
# NOTE: kept in sync with reana_db.utils._add_months /
# get_current_quota_period_start_at_from_anchor. Migrations must be
# self-contained, so any change to the period arithmetic must be applied
# in both places.
```

File: reana_db/models.py [hunk 7]
```diff
@@ -1266,6 +1286,9 @@ class UserResource(Base, Timestamp):
     resource_id = Column(UUIDType, ForeignKey("__reana.resource.id_"), primary_key=True)
     quota_limit = Column(BigInteger())
     quota_used = Column(BigInteger())
+    quota_period_months = Column(Integer, nullable=True)
```
Comment (tiborsimko) [inline reana_db/models.py:1294]: 🤖 **No `CHECK` constraint on `quota_period_months` (LOW)**

The column accepts any integer at the database level, including 0 and negatives. Validation is left entirely to the application layer (`reana_server`). A `CheckConstraint` makes the invariant explicit and protects against manual SQL updates or future callers that bypass the API.

Suggested fix:

```python
quota_period_months = Column(
    Integer,
    CheckConstraint(
        "quota_period_months IS NULL OR quota_period_months > 0",
        name="ck_quota_period_months_positive",
    ),
    nullable=True,
)
```

Note: this also requires adding the constraint in migration `06dbbeef6d9b` (or a follow-up migration), and removing it in `downgrade()`.

File: tests/test_utils.py [hunk 8]
```diff
@@ -91,3 +107,202 @@ def test_get_workflow_with_uuid_or_name(session, new_user):
...
+        (None, 3, datetime(2026, 4, 14, 0, 0, 0), None),
+        (
+            datetime(2026, 1, 15, 12, 0, 0),
+            3,
+            datetime(2026, 5, 20, 0, 0, 0),
+            datetime(2026, 4, 15, 12, 0, 0),
+        ),
+    ],
+)
+def test_get_current_quota_period_start_at_from_anchor(
```
Comment (tiborsimko) [inline tests/test_utils.py]: 🤖 **Missing tests for month-boundary day clamping (NIT)**

`_add_months` (utils.py:525) clamps the day to the last valid day of the target month. This means a user anchored at Jan 31 with a 1-month period gets boundaries Jan 31 → Feb 28 → Mar 28 → ... rather than rolling back to the 31st in months that have 31 days. That is a real semantic choice and the tests should pin it down so a future "simplification" doesn't silently change behaviour.

Suggested fix — add parameterised cases for `_add_months` over `(2026-01-31, 1)`, `(2024-02-29, 12)` (leap year → 2025-02-28), `(2026-12-31, 2)`, and a `get_current_quota_period_start_at_from_anchor` test for a Jan-31 anchor advanced 14 months.

File: reana_db/utils.py [hunk 9]
```diff
@@ -549,20 +658,42 @@ def update_users_cpu_quota(user=None) -> None:
...
+            continue
+
+        _advance_user_cpu_quota_period_if_needed(user_resource_quota)
+
+        quota_period_start_at = _get_current_user_cpu_quota_period_start_at(
+            user_resource_quota
+        )
+
+        if quota_period_start_at:
+            workflows = Session.query(Workflow).filter_by(owner_id=user.id_).all()
```
Comment (tiborsimko) [inline reana_db/utils.py]: 🤖 **Periodic CPU recomputation loads all workflows per user (HIGH)**

For users with periodic CPU accounting enabled, this path loads every `Workflow` row for the owner and then sums clipped CPU time in Python. On populated deployments this can become expensive: the query runs once per user, and unlike `update_workflows_cpu_quota()` just above, it does not defer heavy columns such as logs or workflow specifications. The periodic cron can therefore become much slower and more memory-hungry exactly when it processes users with many historical workflows.

Suggested fix — avoid loading full workflow objects here. At minimum, select only the timestamp columns needed for `_get_accounted_workflow_cpu_milliseconds` or apply the same `defer(Workflow.logs), defer(Workflow.reana_specification)` pattern used in `update_workflows_cpu_quota()`. Longer term, consider calculating the clipped sum in SQL so the periodic path remains one aggregate query per user instead of materialising all workflows in Python.

File: reana_db/config.py [hunk 10]
```diff
@@ -76,6 +76,21 @@
 """Default CPU (in milliseconds) and disk (in bytes) quota limits."""
 
 
+def _get_optional_int_env(var_name):
+    value = os.getenv(var_name)
+    return int(float(value)) if value not in (None, "") else None
```
Comment (tiborsimko) [inline reana_db/config.py]: 🤖 **Default period months parsing accepts surprising values (MEDIUM)**

`_get_optional_int_env()` parses through `float` before converting to `int`. This means values such as `"3.9"` silently become `3`, while `0` or negative values are accepted at config-load time and later disappear through truthiness checks. For an operator-facing Helm/environment setting that controls quota accounting cadence, silently coercing invalid values makes misconfiguration hard to diagnose.

Suggested fix — parse this as a strict positive integer and fail early for invalid values:

```python
def _get_optional_positive_int_env(var_name):
    value = os.getenv(var_name)
    if value in (None, ""):
        return None
    parsed = int(value)
    if parsed <= 0:
        raise ValueError(f"{var_name} must be a positive integer")
    return parsed
```

This also avoids accepting decimal strings whose truncation would be unexpected.

File: reana_db/alembic/versions/20260414_1535_8f2d4e5a4d2c_drop_user_resource_quota_period_anchor.py (hunk 2 above)
Comment (michaelbuchar) [reply reana_db/alembic/versions/20260414_1535_8f2d4e5a4d2c_drop_user_resource_quota_period_anchor.py]: Fixed!

File: reana_db/alembic/versions/20260414_1535_8f2d4e5a4d2c_drop_user_resource_quota_period_anchor.py (hunk 3 above)
Comment (michaelbuchar) [reply reana_db/alembic/versions/20260414_1535_8f2d4e5a4d2c_drop_user_resource_quota_period_anchor.py]: I think it's good to do this optimization, I have implemented this as well!

File: reana_db/utils.py (hunk 4 above)
Comment (michaelbuchar) [reply reana_db/utils.py:594]: I think we should keep this at logging.info since it's good to transmit these messages but let me know if you disagree

File: reana_db/cli.py (hunk 5 above)
Comment (michaelbuchar) [reply reana_db/cli.py]: Fixed!

File: reana_db/alembic/versions/20260414_1535_8f2d4e5a4d2c_drop_user_resource_quota_period_anchor.py (hunk 6 above)
Comment (michaelbuchar) [reply reana_db/alembic/versions/20260414_1535_8f2d4e5a4d2c_drop_user_resource_quota_period_anchor.py]: The duplication is intentional because the migration must remain self-contained and not depend on runtime application code.

File: reana_db/models.py (hunk 7 above)
Comment (michaelbuchar) [reply reana_db/models.py:1294]: Fixed!

File: tests/test_utils.py (hunk 8 above)
Comment (michaelbuchar) [reply tests/test_utils.py]: Fixed!

File: reana_db/utils.py (hunk 9 above)
Comment (michaelbuchar) [reply reana_db/utils.py]: Fixed

File: reana_db/config.py (hunk 10 above)
Comment (michaelbuchar) [reply reana_db/config.py]: Fixed

**Review by tiborsimko — APPROVED** [review]
